### ISSUE TITLE
User authn / authz for silences

### ISSUE THREAD

**Issue by dswarbrick** [prymitive/karma#1347]
This is a request for help / information about whether it is possible to secure certain functionality in Karma, such as the creation of silences.

Assuming that the Alertmanager API has been secured such that POST / DELETE of silences requires authentication, is it possible to configure Karma such that per-user authentication can be applied to the silence create / edit forms, and that authentication passed down to the Alertmanager API?

The problem I'm trying to solve is that in a larger environment with multiple teams responsible for multiple different systems, there is a pretty big security hole in that any authenticated user (or literally anybody, if Alertmanager is not running behind an authenticating reverse proxy) can create silences for services that they technically should not be allowed to. There is quite high potential for the Alertmanager silence API to be abused, e.g. by creating a silence that has sufficiently broad criteria as to silence _all_ alerts.

Are there any suggestions as to how this could possibly be mitigated?

**Comment by prymitive** [comment]
tl;dr - there's no auth or ACLs so you need to provide it in front of both karma and alertmanager

Given that neither karma nor alertmanager has any user management this doesn't seem doable.
I guess one way of implementing this would be to route all traffic from karma to alertmanager via some proxy that validates silence payloads (inspecting matchers to see if nobody is abusing) and compares that to some ACLs based on auth headers (they are passed to alertmanager if sent to karma) OR source IP (if you have a setup with karma instance per team).

**Comment by dswarbrick** [comment]
I found a project that could perhaps be used to mitigate the possibility of a user creating an overly broad silence matcher: https://github.com/openshift/prom-label-proxy

If the alerts already have some kind of "team" label, then multiple prom-label-proxies (with appropriate authn in from of them) could enforce the addition of a "team" matcher in any silences that are created via each prom-label-proxy. At least then, teams are limited to only screwing up their own alerts ;-)

**Comment by prymitive** [comment]
Gonna add some basic authn support via #1461 
Hopefully authz will follow, but that will be a bigger chunk of code I think

**Comment by prymitive** [comment]
- [ ] Authentication support in config & UI
- [ ] Enforcing silence author from credentials when proxy mode is enabled
- [ ] Silence ACL rules in the config
- [ ] Silence ACL checks in the UI
- [ ] Silence ACL checks when proxy mode is enabled

**Comment by prymitive** [comment]
Initial version will have fairly simple ACL rules as regex matchers makes this tricky, will likely need to add config option to disable silence creation with regex matchers, as they could potentially allow bypassing ACL checks.
For example: if I had a rule that blocks silences when `cluster = prod` then someone could create a rule with `cluster =~ p[ro]d` and simple string equality checks wouldn't block it.

**Comment by prymitive** [comment]
There is experimental support in  [v0.56](https://github.com/prymitive/karma/releases/tag/v0.56), full UI support will be tricky here, so right now any blocked silence will just return an error on the submit result page. Might address that in future releases.
